<a href="https://colab.research.google.com/github/yaramirezte/TelecomX_LATAM/blob/main/TelecomX_LATAM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#📌 Extracción

In [23]:
import pandas as pd

In [24]:
url = "https://raw.githubusercontent.com/ingridcristh/challenge2-data-science-LATAM/main/TelecomX_Data.json"

In [25]:
datos_TelecomX = pd.read_json(url)

In [26]:
datos_TelecomX.sample(10)

customerID Churn                                           customer  \
2134  2988-QRAJY    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
5864  8017-UVSZU    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
5453  7471-MQPOS    No  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
3708  5119-KEPFY    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2847  3948-XHGNA   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
4111  5640-CAXOA    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1273  1833-VGRUM   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
3287  4583-PARNH    No  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
6106  8348-HFYIV   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
4819  6599-RCLCJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
2134    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
5864    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
5453    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
3708    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2847     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4111  {'PhoneService': 'No', 'MultipleLines': 'No ph...   
1273    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
3287     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
6106  {'PhoneService': 'No', 'MultipleLines': 'No ph...   
4819    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   

                                               internet  \
2134  {'InternetService': 'Fiber optic', 'OnlineSecu...   
5864  {'InternetService': 'DSL', 'OnlineSecurity': '...   
5453  {'InternetService': 'DSL', 'OnlineSecurity': '...   
3708  {'InternetService': 'Fiber optic', 'OnlineSecu...   
2847  {'InternetService': 'DSL', 'OnlineSecurity': '...   
4111  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1273  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3287  {'InternetService': 'Fiber optic', 'OnlineSecu...   
6106  {'InternetService': 'DSL', 'OnlineSecurity': '...   
4819  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                                account  
2134  {'Contract': 'Two year', 'PaperlessBilling': '...  
5864  {'Contract': 'Month-to-month', 'PaperlessBilli...  
5453  {'Contract': 'One year', 'PaperlessBilling': '...  
3708  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2847  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4111  {'Contract': 'One year', 'PaperlessBilling': '...  
1273  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3287  {'Contract': 'Month-to-month', 'PaperlessBilli...  
6106  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4819  {'Contract': 'Month-to-month', 'PaperlessBilli...

#🔧 Transformación

In [27]:
datos_TelecomX['customer'].iloc[0]

{'gender': 'Female',
 'SeniorCitizen': 0,
 'Partner': 'Yes',
 'Dependents': 'Yes',
 'tenure': 9}

El dataset presenta una estructura JSON anidada, donde varias columnas contienen diccionarios con información detallada del cliente, servicios contratados y datos de facturación.

In [28]:
datos_TelecomX.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


In [29]:
datos_TelecomX.dtypes

customerID    object
Churn         object
customer      object
phone         object
internet      object
account       object
dtype: object

In [30]:
datos_TelecomX.describe()

customerID Churn                                           customer  \
count         7267  7267                                               7267   
unique        7267     3                                                891   
top     9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
freq             1  5174                                                223   

                                                 phone  \
count                                             7267   
unique                                               3   
top     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
freq                                              3495   

                                                 internet  \
count                                                7267   
unique                                                129   
top     {'InternetService': 'No', 'OnlineSecurity': 'N...   
freq                                                 1581   

                                                  account  
count                                                7267  
unique                                               6931  
top     {'Contract': 'Month-to-month', 'PaperlessBilli...  
freq                                                    6

In [32]:
datos_TelecomX.isnull().sum()

customerID    0
Churn         0
customer      0
phone         0
internet      0
account       0
dtype: int64

In [33]:
datos_TelecomX.isna().sum()

customerID    0
Churn         0
customer      0
phone         0
internet      0
account       0
dtype: int64

In [34]:
datos_TelecomX.dtypes

customerID    object
Churn         object
customer      object
phone         object
internet      object
account       object
dtype: object

In [35]:
datos_TelecomX.map(type).head()

customerID          Churn        customer           phone  \
0  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   
1  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   
2  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   
3  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   
4  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   

         internet         account  
0  <class 'dict'>  <class 'dict'>  
1  <class 'dict'>  <class 'dict'>  
2  <class 'dict'>  <class 'dict'>  
3  <class 'dict'>  <class 'dict'>  
4  <class 'dict'>  <class 'dict'>

In [36]:
# pd.json_normalize() Recorre cada diccionario, entra dentro de los diccionarios anidados, saca cada clave interna, crea columnas nuevas usando notación con puntos
# datos_TelecomX.to_dict(orient="records") - Convierte el DataFrame en una lista de diccionarios, uno por fila.
datos_TelecomX_normalizados = pd.json_normalize(datos_TelecomX.to_dict(orient="records"))

In [41]:
datos_TelecomX_normalizados.sample(5)

customerID Churn customer.gender  customer.SeniorCitizen  \
2573  3587-PMCOY    No            Male                       0   
1702  2410-CIYFZ    No            Male                       0   
7213  9921-QFQUL    No          Female                       0   
5236  7138-GIRSH    No            Male                       0   
186   0295-QVKPB    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
2573               No                  No               10                Yes   
1702               No                 Yes                2                Yes   
7213              Yes                  No               23                Yes   
5236               No                  No               50                Yes   
186                No                  No                5                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
2573                  No              Fiber optic  ...                    No   
1702                  No                       No  ...   No internet service   
7213                 Yes              Fiber optic  ...                    No   
5236                 Yes                      DSL  ...                    No   
186                   No                      DSL  ...                    No   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
2573                       Yes                   No                  Yes   
1702       No internet service  No internet service  No internet service   
7213                       Yes                   No                   No   
5236                       Yes                  Yes                   No   
186                        Yes                  Yes                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
2573                      Yes         One year                      Yes   
1702      No internet service   Month-to-month                       No   
7213                      Yes   Month-to-month                       No   
5236                      Yes         Two year                       No   
186                        No   Month-to-month                      Yes   

          account.PaymentMethod account.Charges.Monthly  account.Charges.Total  
2573           Electronic check                   98.90                1064.95  
1702               Mailed check                   20.40                   42.9  
7213               Mailed check                   90.15                2044.95  
5236  Bank transfer (automatic)                   69.50                 3418.2  
186     Credit card (automatic)                   63.95                  318.1  

[5 rows x 21 columns]

In [42]:
datos_TelecomX_normalizados.duplicated().sum()

np.int64(0)

In [43]:
datos_TelecomX_normalizados['account.Contract'].unique()

array(['One year', 'Month-to-month', 'Two year'], dtype=object)

In [45]:
datos_TelecomX_normalizados['internet.InternetService'].unique()

array(['DSL', 'Fiber optic', 'No'], dtype=object)

In [46]:
datos_TelecomX_normalizados['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

In [47]:
datos_TelecomX_normalizados.columns

Index(['customerID', 'Churn', 'customer.gender', 'customer.SeniorCitizen',
       'customer.Partner', 'customer.Dependents', 'customer.tenure',
       'phone.PhoneService', 'phone.MultipleLines', 'internet.InternetService',
       'internet.OnlineSecurity', 'internet.OnlineBackup',
       'internet.DeviceProtection', 'internet.TechSupport',
       'internet.StreamingTV', 'internet.StreamingMovies', 'account.Contract',
       'account.PaperlessBilling', 'account.PaymentMethod',
       'account.Charges.Monthly', 'account.Charges.Total'],
      dtype='object')

In [48]:
datos_TelecomX_normalizados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

In [49]:
datos_TelecomX_normalizados.dtypes

customerID                    object
Churn                         object
customer.gender               object
customer.SeniorCitizen         int64
customer.Partner              object
customer.Dependents           object
customer.tenure                int64
phone.PhoneService            object
phone.MultipleLines           object
internet.InternetService      object
internet.OnlineSecurity       object
internet.OnlineBackup         object
internet.DeviceProtection     object
internet.TechSupport          object
internet.StreamingTV          object
internet.StreamingMovies      object
account.Contract              object
account.PaperlessBilling      object
account.PaymentMethod         object
account.Charges.Monthly      float64
account.Charges.Total         object
dtype: object

El conjunto de datos presenta información demográfica de los clientes características de los servicios contratados y datos contractuales y financieros. Tambien se consultó el diccionario de datos proporcionado. A partir de esta revisión, se identificó la variable Churn como objetivo, y se seleccionaron como variables relevantes aquellas relacionadas con el perfil del cliente, los servicios contratados y las condiciones del contrato.

In [50]:
df = datos_TelecomX_normalizados.copy()

In [51]:
for col in df.select_dtypes(include="object").columns:
  df[col] = df[col].str.strip()

In [55]:
df['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

In [97]:
(df['Churn'] == '').sum()

np.int64(224)

In [98]:
churn_vacios = df[df['Churn'] == '']

In [99]:
churn_vacios

customerID Churn customer.gender  customer.SeniorCitizen  \
30    0047-ZHDTW                Female                       0   
75    0120-YZLQA                  Male                       0   
96    0154-QYHJU                  Male                       0   
98    0162-RZGMZ                Female                       1   
175   0274-VVQOQ                  Male                       1   
...          ...   ...             ...                     ...   
7158  9840-GSRFX                Female                       0   
7180  9872-RZQQB                Female                       0   
7211  9920-GNDMB                  Male                       0   
7239  9955-RVWSC                Female                       0   
7247  9966-VYRTZ                Female                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
30                 No                  No               11                Yes   
75                 No                  No               71                Yes   
96                 No                  No               29                Yes   
98                 No                  No                5                Yes   
175               Yes                  No               65                Yes   
...               ...                 ...              ...                ...   
7158               No                  No               14                Yes   
7180              Yes                  No               49                 No   
7211               No                  No                9                Yes   
7239              Yes                 Yes               67                Yes   
7247              Yes                 Yes               31                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
30                   Yes              Fiber optic  ...                    No   
75                    No                       No  ...   No internet service   
96                    No                      DSL  ...                   Yes   
98                    No                      DSL  ...                   Yes   
175                  Yes              Fiber optic  ...                   Yes   
...                  ...                      ...  ...                   ...   
7158                 Yes                      DSL  ...                   Yes   
7180    No phone service                      DSL  ...                    No   
7211                 Yes              Fiber optic  ...                    No   
7239                  No                       No  ...   No internet service   
7247                  No                       No  ...   No internet service   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
30                          No                   No                   No   
75         No internet service  No internet service  No internet service   
96                          No                  Yes                   No   
98                          No                  Yes                   No   
175                        Yes                   No                  Yes   
...                        ...                  ...                  ...   
7158                        No                   No                   No   
7180                        No                   No                  Yes   
7211                        No                   No                   No   
7239       No internet service  No internet service  No internet service   
7247       No internet service  No internet service  No internet service   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
30                         No   Month-to-month                      Yes   
75        No internet service         Two year                      Yes   
96                         No         One year                      Yes   
98                         No   Month-to-month                    

In [100]:
churn_vacios[['customer.tenure', 'account.Charges.Total', 'account.Charges.Monthly']].head()

customer.tenure  account.Charges.Total  account.Charges.Monthly
30                11                 929.30                    79.00
75                71                1355.10                    19.90
96                29                1696.20                    58.75
98                 5                 287.85                    59.90
175               65                6792.45                   103.15

In [101]:
# eliminar los registros con df['Churn'] vacios
df = df[df['Churn'] != '']

In [56]:
# df['Churn'] = df['Churn'].str.capitalize()

In [103]:
df['Churn'].unique()

array(['No', 'Yes'], dtype=object)

Se identificaron 224 registros con la variable objetivo Churn vacía. Al explorar estos clientes se observó que poseen historial de facturación y antigüedad variable (customer.tenure), pero no es posible inferir su estado de evasión. Por este motivo, estos registros fueron eliminados del conjunto de datos, asegurando la integridad de la variable objetivo.

In [77]:
df['account.Charges.Total'].dtype

dtype('O')

In [79]:
df['account.Charges.Total']= pd.to_numeric(df['account.Charges.Total'], errors='coerce')

In [81]:
df['account.Charges.Total'].dtype

dtype('float64')

In [82]:
df.isnull().sum()

customerID                    0
Churn                         0
customer.gender               0
customer.SeniorCitizen        0
customer.Partner              0
customer.Dependents           0
customer.tenure               0
phone.PhoneService            0
phone.MultipleLines           0
internet.InternetService      0
internet.OnlineSecurity       0
internet.OnlineBackup         0
internet.DeviceProtection     0
internet.TechSupport          0
internet.StreamingTV          0
internet.StreamingMovies      0
account.Contract              0
account.PaperlessBilling      0
account.PaymentMethod         0
account.Charges.Monthly       0
account.Charges.Total        11
dtype: int64

In [84]:
df.loc[df['account.Charges.Total'].isnull(), ['account.Charges.Total', 'customer.tenure', 'account.Charges.Monthly']]

account.Charges.Total  customer.tenure  account.Charges.Monthly
975                     NaN                0                    56.05
1775                    NaN                0                    20.00
1955                    NaN                0                    61.90
2075                    NaN                0                    19.70
2232                    NaN                0                    20.25
2308                    NaN                0                    25.35
2930                    NaN                0                    73.35
3134                    NaN                0                    25.75
3203                    NaN                0                    52.55
4169                    NaN                0                    80.85
5599                    NaN                0                    19.85

In [85]:
df.loc[df['customer.tenure'] == 0, 'account.Charges.Total'] = 0

In [87]:
df['account.Charges.Total'].isnull().sum()

np.int64(0)

Se identificaron 11 valores nulos en la variable account.Charges.Total. Al analizar estos registros se comprobó que corresponden a clientes con antigüedad (customer.tenure) igual a cero, quienes aún no han generado cargos acumulados. Por este motivo, los valores nulos fueron actualizados con el valor 0, garantizando la coherencia de la información financiera.

In [88]:
df.sample(5)

customerID Churn customer.gender  customer.SeniorCitizen  \
6366  8749-JMNKX   Yes            Male                       1   
2554  3566-VVORZ    No          Female                       0   
1094  1550-EENBN    No          Female                       0   
5234  7137-NAXML    No            Male                       0   
2273  3170-NMYVV    No          Female                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
6366              Yes                  No               51                Yes   
2554              Yes                  No               12                Yes   
1094               No                  No               31                Yes   
5234               No                  No               23                Yes   
2273              Yes                 Yes               50                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
6366                 Yes              Fiber optic  ...                    No   
2554                  No                      DSL  ...                    No   
1094                 Yes              Fiber optic  ...                   Yes   
5234                  No                       No  ...   No internet service   
2273                  No                       No  ...   No internet service   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
6366                       Yes                   No                  Yes   
2554                        No                   No                   No   
1094                        No                   No                  Yes   
5234       No internet service  No internet service  No internet service   
2273       No internet service  No internet service  No internet service   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
6366                       No   Month-to-month                      Yes   
2554                       No   Month-to-month                       No   
1094                      Yes   Month-to-month                       No   
5234      No internet service         One year                       No   
2273      No internet service         Two year                       No   

          account.PaymentMethod account.Charges.Monthly  account.Charges.Total  
6366  Bank transfer (automatic)                   93.80                4750.95  
2554  Bank transfer (automatic)                   45.90                 505.95  
1094           Electronic check                  104.35                3205.60  
5234  Bank transfer (automatic)                   20.15                 405.60  
2273  Bank transfer (automatic)                   20.15                 930.90  

[5 rows x 21 columns]

In [90]:
df['internet.InternetService'].unique()

array(['DSL', 'Fiber optic', 'No'], dtype=object)

In [92]:
df['internet.OnlineSecurity'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [93]:
# Consultar
df.loc[df['internet.InternetService'] == 'No',['internet.InternetService', 'internet.OnlineSecurity', 'internet.OnlineBackup']]


internet.InternetService internet.OnlineSecurity internet.OnlineBackup
20                         No     No internet service   No internet service
23                         No     No internet service   No internet service
24                         No     No internet service   No internet service
27                         No     No internet service   No internet service
28                         No     No internet service   No internet service
...                       ...                     ...                   ...
7250                       No     No internet service   No internet service
7252                       No     No internet service   No internet service
7256                       No     No internet service   No internet service
7257                       No     No internet service   No internet service
7261                       No     No internet service   No internet service

[1581 rows x 3 columns]

Se verificaron las variables relacionadas con servicios de internet y se observó que, para los clientes sin servicio de internet (InternetService = "No"), las variables de servicios adicionales toman el valor "No internet service". Esta codificación es consistente y representa correctamente la ausencia de dichos servicios, por lo que no se realizaron modificaciones en estas columnas.

In [94]:
df.duplicated().sum()

np.int64(0)

In [95]:
df.info()
df.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

customerID Churn customer.gender  customer.SeniorCitizen  \
count         7267  7267            7267             7267.000000   
unique        7267     3               2                     NaN   
top     9995-HOTOH    No            Male                     NaN   
freq             1  5174            3675                     NaN   
mean           NaN   NaN             NaN                0.162653   
std            NaN   NaN             NaN                0.369074   
min            NaN   NaN             NaN                0.000000   
25%            NaN   NaN             NaN                0.000000   
50%            NaN   NaN             NaN                0.000000   
75%            NaN   NaN             NaN                0.000000   
max            NaN   NaN             NaN                1.000000   

       customer.Partner customer.Dependents  customer.tenure  \
count              7267                7267      7267.000000   
unique                2                   2              NaN   
top                  No                  No              NaN   
freq               3749                5086              NaN   
mean                NaN                 NaN        32.346498   
std                 NaN                 NaN        24.571773   
min                 NaN                 NaN         0.000000   
25%                 NaN                 NaN         9.000000   
50%                 NaN                 NaN        29.000000   
75%                 NaN                 NaN        55.000000   
max                 NaN                 NaN        72.000000   

       phone.PhoneService phone.MultipleLines internet.InternetService  ...  \
count                7267                7267                     7267  ...   
unique                  2                   3                        3  ...   
top                   Yes                  No              Fiber optic  ...   
freq                 6560                3495                     3198  ...   
mean                  NaN                 NaN                      NaN  ...   
std                   NaN                 NaN                      NaN  ...   
min                   NaN                 NaN                      NaN  ...   
25%                   NaN                 NaN                      NaN  ...   
50%                   NaN                 NaN                      NaN  ...   
75%                   NaN                 NaN                      NaN  ...   
max                   NaN                 NaN                      NaN  ...   

       internet.OnlineBackup internet.DeviceProtection internet.TechSupport  \
count                   7267                      7267                 7267   
unique                     3                         3                    3   
top                       No                        No                   No   
freq                    3182                      3195                 3582   
mean                     NaN                       NaN                  NaN   
std                      NaN                       NaN                  NaN   
min                      NaN                       NaN                  NaN   
25%                      NaN                       NaN                  NaN   
50%                      NaN                       NaN                  NaN   
75%                      NaN                       NaN                  NaN   
max                      NaN                       NaN                  NaN   

       internet.StreamingTV internet.StreamingMovies account.Contract  \
count                  7267                     7267             7267   
unique                    3                        3                3   
top                      No                       No   Month-to-month   
freq                   2896                     2870             4005   
mean                    NaN                      NaN              NaN   
std                     NaN                      NaN              NaN   
min          

In [96]:
for col in df.select_dtypes(include='object').columns:
    print(col, "→", df[col].unique(), "\n")

customerID → ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH'] 

Churn → ['No' 'Yes' ''] 

customer.gender → ['Female' 'Male'] 

customer.Partner → ['Yes' 'No'] 

customer.Dependents → ['Yes' 'No'] 

phone.PhoneService → ['Yes' 'No'] 

phone.MultipleLines → ['No' 'Yes' 'No phone service'] 

internet.InternetService → ['DSL' 'Fiber optic' 'No'] 

internet.OnlineSecurity → ['No' 'Yes' 'No internet service'] 

internet.OnlineBackup → ['Yes' 'No' 'No internet service'] 

internet.DeviceProtection → ['No' 'Yes' 'No internet service'] 

internet.TechSupport → ['Yes' 'No' 'No internet service'] 

internet.StreamingTV → ['Yes' 'No' 'No internet service'] 

internet.StreamingMovies → ['No' 'Yes' 'No internet service'] 

account.Contract → ['One year' 'Month-to-month' 'Two year'] 

account.PaperlessBilling → ['Yes' 'No'] 

account.PaymentMethod → ['Mailed check' 'Electronic check' 'Credit card (automatic)'
 'Bank transfer (automatic)'] 



#📊 Carga y análisis

#📄Informe final